src/handler/httpfuture.rs

```diff
@@ -0,0 +1,65 @@
+use futures::channel::mpsc;
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use crate::error::Result;
+use crate::handler::commandfuture::CommandFuture;
+use crate::handler::http::HttpRequest;
+use crate::handler::navigationfuture::NavigationFuture;
+use crate::handler::target::TargetMessage;
+use chromiumoxide_types::Command;
+
+type ArcRequest = Option<Arc<HttpRequest>>;
+
+pin_project! {
+    pub struct HttpFuture<T: Command> {
+        #[pin]
+        command: CommandFuture<T>,
+        #[pin]
+        navigation: NavigationFuture,
+
+        command_complete: bool,
+    }
+}
+
+impl<T: Command> HttpFuture<T> {
+    pub fn new(sender: mpsc::Sender<TargetMessage>, command: CommandFuture<T>) -> Self {
+        Self {
+            command,
+            navigation: NavigationFuture::new(sender),
+            command_complete: false,
+        }
+    }
+}
+
+impl<T> Future for HttpFuture<T>
+where
+    T: Command,
+{
+    type Output = Result<ArcRequest>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        // 1. First complete command request
+        // 2. After command request is complete flip a switch of `command_complete` field
+        // 3. Switch polls navigation
+        if *this.command_complete {
+            this.navigation.poll(cx)
+        } else {
+            match this.command.poll(cx) {
+                Poll::Ready(Ok(_command_response)) => {
+                    *this.command_complete = true;
+                }
+                Poll::Ready(Err(e)) => return Poll::Ready(Err(e.into())),
+                Poll::Pending => return Poll::Pending,
+            }
+
+            cx.waker().wake_by_ref();
```

Comment:
why do we need to manually wake here?

Reply:
Manual wake allows to switch faster to polling NavigationFuture. 
Poll::Ready(Err(_)) and Poll::Pending branches have explicit return statements. Rewrote this a little, now it is more readable to my eye.

```
            match this.command.poll(cx) {
                Poll::Ready(Ok(_command_response)) => {
                    *this.command_complete = true;

                    cx.waker().wake_by_ref();
                    Poll::Pending
                }
                Poll::Ready(Err(e)) => Poll::Ready(Err(e.into())),
                Poll::Pending => Poll::Pending,
            }
```